### PR TITLE
[Breakfast] Scrabble

### PR DESCRIPTION
**Pull request by catchingash**
Meets all breakfast requirements + handles inappropriate input. Includes 15 passing specs. Handles gets.chomp in rspec so that the user does not have to enter appropriate input when running the specs.

Requires 'pathname'

### CODE REVIEW

File: lib/scrabble.rb [hunk 1]
```diff
@@ -1,4 +1,8 @@
 module Scrabble
   # require your gems and classes here
-  # require_relative 'scrabble/whatevs'
+  require 'pathname'
+
+  ROOT_DIR = Pathname(__dir__).parent ## QUESTION: Is Pathname(__dir__) == Pathname.new(__dir__) ?
```
Comment (mhotan) [inline lib/scrabble.rb:5]: Ruby doc was inconclusive.  Not sure if it appears to work the same way then go for it.  I feel like for the current purpose it will very quickly resolve in error if it doesn't.

File: lib/scrabble/scrabble.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    # checks word for not-allowed content. If contains, prompts for new input until accepted.
+    def self.acceptable_input_check(word)
+      # checks for nil, special characters. NOTE: '' is treated as zero, and is an accepted input.
+      status = (
+      if word.is_a?(String) # checks for nil input, arrays, integers (not in a string), etc.
+        word.chars.each do |letter|
+          return false unless ('a'..'z').include?(letter) || ('A'..'Z').include?(letter) #.downcase doesn't work on integers
+        end
+        word.length <= 7
```
Comment (mhotan) [inline lib/scrabble/scrabble.rb:32]: What is the purpose of this preposition?

File: lib/scrabble.rb (hunk 1 above)
Comment (catchingash) [inline lib/scrabble.rb:5]: When I tried it, it worked the same way. I just wasn't sure if there was a preferred way, or if there are certain situations where it wouldn't work the same.

File: lib/scrabble/scrabble.rb (hunk 2 above)
Comment (catchingash) [inline lib/scrabble/scrabble.rb:32]: word.length <= 7? Scrabble doesn't allow for words longer than 7 characters, so this was  to prevent 8+ characters.

Comment (mhotan) [inline lib/scrabble/scrabble.rb:32]: Right I see, that makes sense thanks for the clarification

File: lib/scrabble/scrabble.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      f: 4, h: 4, v: 4, w: 4, y: 4,
+      k: 5,
+      j: 8, x: 8,
+      q: 10, z: 10
+    }
+
+    # .score(word): returns the total score value for the given word [breakfast]
+    def self.score(word) # word is input as a string (case insensitive)
+      letters = convert_input(word)
+      letters.inject(0){ |sum, letter| sum + VALUES[letter.to_sym] }
```
Comment (mhotan) [inline lib/scrabble/scrabble.rb:15]: This is beautiful.  Great use of inject.  Short, concise functionality.

Comment (catchingash) [inline lib/scrabble/scrabble.rb:15]: Thanks! I discovered inject/reduce and was dying to use it!!

File: lib/scrabble/scrabble.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        if shortest_tied_words.length == 1 # if true, there is only one shortest word
+          shortest_tied_words[0]
+        else # there are ties for shortest word with highest score
+          find_first_in_original(original_words, shortest_tied_words)
+        end
+      end
+    end
+
+    # [breakfast]: If the there are multiple words that are the same score and same length, pick the first one in supplied list
+    def self.find_first_in_original(original_words, tied_words)
```
Comment (mhotan) [inline lib/scrabble/scrabble.rb:83]: I like the modularization of different methods.  Each one has a clear purpose.  Well done.  I am trying to think of a clean way to prevent having to pass a reference to the original array to each method.  Nothing comes to mind with the way your current code is structured.  Looks good.

File: spec/lib/scrabble_spec.rb [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    #first = 8
+    #seven = 8
+    #doll = 5
+    #all = 3
+
+    it "returns the word with the highest score" do
+      expect(Scrabble::Scrabble.highest_score_from(['zz', 'aa', 'cat', 'dog'].shuffle)).to eq('zz')
+    end
+
+    context "when words tie" do
```
Comment (mhotan) [inline spec/lib/scrabble_spec.rb:78]: Really solid tests.  Way to test the boundary cases. :+1:

File: spec/lib/scrabble_spec.rb [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  describe ".acceptable_input_check" do
+    before do
+      $stdin = StringIO.new("cat\n")
+    end
+
+    after do
+      $stdin = STDIN
+    end
+
+    it "does not accept nil" do
```
Comment (mhotan) [inline spec/lib/scrabble_spec.rb:18]: Way to think about testing input.  This helps a newcomer to your code understand how the methods should behave.

**Comment by mhotan** [comment]
A very well thought through test suite.  The first thing I looked at was the tests.  Tests provide me an insight into how you thought about and approached the problem.  I liked how thorough and clear they were.  Good use of Class and enumerable methods.  Good use of the Hash data structure.  Overrall great job.

File: lib/scrabble/scrabble.rb (hunk 4 above)
Comment (catchingash) [inline lib/scrabble/scrabble.rb:83]: How would you suggest restructuring?

As a note: with how Ruby's group_by (currently) works, each element of tied_words is still in the same order as the original... but since this isn't guaranteed in the docs, I built extra functionality to check for it.

**Comment by catchingash** [comment]
Thanks :) 

I really appreciate the time you took to review this!
